core/src/main/scala/nl/vroste/zio/kinesis/client/zionative/leasecoordinator/DefaultLeaseCoordinator.scala

```diff
@@ -325,33 +326,45 @@ private class DefaultLeaseCoordinator(
    */
   val takeLeases: ZIO[Any, Throwable, Unit] =
     for {
-      leases          <- state.get.map(_.currentLeases.values.toSet)
-      shards          <- state.get.map(_.shards.view.map { case (k, v) => ShardId.unwrap(k) -> v }.toMap)
+      _      <- currentShards.flatMap(updateShards)
+      leases <- state.get.map(_.currentLeases.values.toSet)
+      shards <- state.get.map(_.shards.view.map { case (k, v) => ShardId.unwrap(k) -> v }.toMap)
+
       leaseMap         = leases.map(s => s.lease.key -> s.lease).toMap
       openLeases       = leases.collect {
                            case LeaseState(lease, completed @ _, lastUpdated) if !shardHasEnded(lease) =>
                              (lease, lastUpdated)
                          }
       consumableShards = shardsReadyToConsume(shards, leaseMap)
-      desiredShards   <- strategy.desiredShards(openLeases, consumableShards.keySet, workerId)
-      _               <- ZIO.logInfo(s"Desired shard assignment: ${desiredShards.mkString(",")}")
-      toTake           = leaseMap.values.filter(l => desiredShards.contains(l.key) && !l.owner.contains(workerId))
-      _               <- claimLeasesForShardsWithoutLease(desiredShards, shards, leaseMap)
-      _               <- ZIO
-                           .logInfo(s"Going to take ${toTake.size} leases from other workers: ${toTake.mkString(",")}")
-                           .when(toTake.nonEmpty)
-      _               <- foreachParNUninterrupted_(settings.maxParallelLeaseAcquisitions)(toTake) { lease =>
-                           val updatedLease = lease.claim(workerId)
-                           (table.claimLease(applicationName, updatedLease) *>
-                             serialExecutionByShard(updatedLease.key)(registerNewAcquiredLease(updatedLease))).catchAll {
-                             case Right(UnableToClaimLease) =>
-                               ZIO.logInfo(
-                                 s"Unable to claim lease for shard ${lease.key}, beaten to it by another worker?"
-                               )
-                             case Left(e)                   =>
-                               ZIO.logError(s"Got error ${e}") *> ZIO.fail(e)
-                           }
+
+      // before we consume a shard with a missing parent, we need to ensure we didn't just miss it due to a race
```

Comment:
This can happen in two cases:

1. We startup with an existing leasemap and hit the end of a shard before a single listShards in the consumer has completed. In this case shards is empty and we will start consuming the childshards without waiting for any other parent shards to complete.
2. We have multiple levels of empy / small childshards and finish consumption before we become aware of any other parents at that level. This might look something like this (K = known shard; F = known, finished shard; U = unknown shard):

```
1F - 3K
  \
   4F - 7K
        /
2F - 5U
  \
   6U
 ```
 
 Here we start consuming shard 1 while another consumer is responsible for 2. As we keep finding child shards, we eventually come into a situation in which can start to consume 7, eventhough the other consumer might not have finished its other parent 5 yet.